restapi/api/server.go

```diff
@@ -0,0 +1,88 @@
+package api
+
+import (
+	"github.com/gin-gonic/gin"
+	"github.com/sirupsen/logrus"
+	"github.com/swaggo/files"
+	"github.com/swaggo/gin-swagger"
+	"io"
+	"net/http"
+	"os"
+)
+
+//limitNumClients uses a go channel to rate limit a handler.
+// It is a golang buffered channel so you can put maxClients empty structs
+//into the channel without blocking. The maxClients+1 invocation will block
+//until another handler finishes and removes one empty struct from the channel.
+func limitNumClients(f http.HandlerFunc, maxClients int) http.HandlerFunc {
+	sema := make(chan struct{}, maxClients)
+
+	return func(w http.ResponseWriter, req *http.Request) {
+		sema <- struct{}{}
+		defer func() { <-sema }()
+		f(w, req)
+	}
+}
+
+/*
+func limitNumClientsUDID(f http.HandlerFunc) http.HandlerFunc {
+	maxClients := 1
+	semaMap := map[string]chan struct{}{}
+	mux := sync.Mutex{}
+	return func(w http.ResponseWriter, r *http.Request) {
+		udid := strings.TrimSpace(r.URL.Query().Get("udid"))
+		if udid == "" {
+			serverError("missing udid", http.StatusBadRequest, w)
+			return
+		}
+		mux.Lock()
+		var sema chan struct{}
+		sema, ok := semaMap[udid]
+		if !ok {
+			sema = make(chan struct{}, maxClients)
+			semaMap[udid] = sema
+		}
+		mux.Unlock()
+		sema <- struct{}{}
+		defer func() { <-sema }()
+		f(w, r)
+	}
+}
+*/
+
+func Main() {
+	router := gin.Default()
+	log := logrus.New()
+	myfile, _ := os.Create("go-ios.log")
+	gin.DefaultWriter = io.MultiWriter(myfile, os.Stdout)
+	// Add event-streaming headers
+	router.Use(MyLogger(log), gin.Recovery())
+	router.Use(HeadersMiddleware())
+
+	v1 := router.Group("/api/v1", gin.BasicAuth(gin.Accounts{
+		"admin": "admin123", // username : admin, password : admin123
+	}))
+	device := v1.Group("/device/:udid")
+	device.GET("/info", Info)
+	device.GET("/screenshot", Screenshot)
+
+	v1.GET("/listen", Listen)
+	v1.GET("/list", List)
+
+	router.GET("/swagger/*any", ginSwagger.WrapHandler(swaggerFiles.Handler))
+
+	err := router.Run(":8080")
```

Comment:
We should provide an option for user selected port, maybe using `flag`, `8080` is quite generic and could already be used by something else

Reply:
agreed.
I have not yet thought about how to manage configuration. But I think also other things will need to be configured later on. One good example is rate limiting. Some endpoints will have to be rate limited so they cannot be executed concurrently, or only with a limit. F.ex. hitting the screenshot endpoint 20 times in parallel will probably kill it, or we cache it somehow.